ehrql/renderers.py

```diff
@@ -1,134 +1,75 @@
-        rows.append(f"<tr>{row}</tr>")
-    rows = "".join(rows)
-
-    return f"{START_MARKER}<table><thead>{headers}</thead><tbody>{rows}</tbody></table>{END_MARKER}"
-
-
-def records_to_ascii_table(records: list[dict]):
-    width = 17
-    lines = []
-    headers_written = False
-    for record in records:
-        if not headers_written:
-            lines.append(" | ".join(name.ljust(width) for name in record.keys()))
-            lines.append("-+-".join("-" * width for _ in record.keys()))
-            headers_written = True
-        lines.append(" | ".join(str(value).ljust(width) for value in record.values()))
-    return "\n".join(line.strip() for line in lines) + "\n"
-
-
-def _truncate_html_table(table_repr: str, head: int | None, tail: int | None):
-    """
-    Truncate an html table string to the first/last N rows, with a row of ...
-    values to indicate where it's been truncated
-    """
-    regex = re.compile(
-        rf"(?P<start>^{START_MARKER}<table>.*<tbody>)(?P<rows><tr>.*<\/tr>)(?P<end><\/tbody>.*<\/table>{END_MARKER})"
-    )
-    match = regex.match(table_repr)
-    if match is None:
-        # if we can't parse the table, return None and let the fallback handle it
-        return
-
-    start = match.group("start")
-    rows = match.group("rows")
-    end = match.group("end")
+def headtail(sequence: typing.Sequence, head: int = 0, tail: int = 0):
+    if head == tail == 0:
+        return sequence, False, []
 
     # Check we have enough rows to truncate to head and tail
-    if rows.count("<tr>") <= ((head or 0) + (tail or 0)):
-        return table_repr
+    if len(sequence) <= head + tail:
+        return sequence, False, []
 
-    # split row on <tr> tokens, remove any empty strings (this will lose the first <tr> of
-    # each row, but we'll add it in again later)
-    rows = [row for row in rows.split("<tr>") if row]
-    # compose an "ellipsis row" to mark the place of truncated rows
-    td_count = rows[0].count("<td>")
-    ellipsis_row = f"{'<td>...</td>' * td_count}</tr>"
+    head_rows = []
+    tail_rows = []
 
-    # Build the list of rows we need to include, with ellipsis rows where necessary
-    truncated_rows = []
+    if head:
+        head_rows = sequence[:head]
 
-    head_rows = rows[:head] if head is not None else [ellipsis_row]
-    truncated_rows.extend(head_rows)
+    if tail:
+        tail_rows = sequence[-tail:]
 
-    if head is not None and tail is not None:
-        truncated_rows.append(ellipsis_row)
+    return head_rows, True, tail_rows
 
-    tail_rows = rows[-tail:] if tail is not None else [ellipsis_row]
-    truncated_rows.extend(tail_rows)
 
-    # re-join the truncated rows with <tr>
-    truncated_rows = "<tr>" + "<tr>".join(truncated_rows)
-    return start + truncated_rows + end
+def records_to_html_table(records: list[dict], head: int = 0, tail: int = 0):
+    rows = []
+    headers = "".join([f"<th>{header}</th>" for header in records[0].keys()])
```

Comment:
Could records ever be an empty list? If so then I think this throws an error now which wouldn't have happened before.

Reply:
Good catch, will fix.  It might be better to supply the header info independently of the rows, so we can show headers for an empty table, but I am not sure how to do that, so I will go for just not breaking.